hw/ip/otbn/util/shared/control_flow.py

```diff
@@ -0,0 +1,392 @@
+                if isinstance(loc, LoopStart):
+                    # Prevent loops in printing by checking we didn't already print this loop
+                    if loc.loop_start_pc not in already_printed:
+                        out.extend([(indent + child_indent, line)
+                                    for indent, line in self._pretty_lines(
+                                        program, loc.loop_start_pc, concise,
+                                        already_printed, call_stack)])
+            else:
+                insn = program.get_insn(sec.end)
+                operands = program.get_operands(sec.end)
+                if insn.mnemonic == 'jal' and operands['grd'] == 1:
+                    # push to to the call stack
+                    call_stack = [sec.end + 4] + call_stack
+                    child_indent += 2
+                out.extend([
+                    (indent + child_indent, line)
+                    for indent, line in self._pretty_lines(
+                        program, loc.pc, concise, already_printed, call_stack)
+                ])
+        return out
+
+    def pretty(self, program, indent=0, concise=False):
+        '''Returns string for pretty-printing.'''
+        lines = self._pretty_lines(program, self.start, concise, set())
+        lines = [(max(0, indent + lindent), line) for (lindent, line) in lines]
+        lines = [
+            '{}{}'.format(' ' * lindent, line) for (lindent, line) in lines
+        ]
+        return '\n'.join(lines)
+
+
+def _clip_to_32(value: int) -> int:
+    '''Truncates a number to 32 bits.'''
+    return value & ((1 << 32) - 1)
+
+
+def _get_next_control_locations(insn: Insn, operands: Dict[str, int],
+                                pc: int) -> List[ControlLoc]:
+    '''Given a control-flow instruction, returns the possible destinations.
+
+    Raises a RuntimeError if the instruction is not a recognized control-flow
+    instruction, or if some unsupported construct is encountered.
+    '''
+    if insn.mnemonic == 'beq' or insn.mnemonic == 'bne':
+        branch_dst = _clip_to_32(operands['offset'])
+        # Branch can either go to branch_dst or next PC
+        return [ControlLoc(branch_dst), ControlLoc(pc + 4)]
+    elif insn.mnemonic == 'jal':
+        jump_dst = _clip_to_32(operands['offset'])
+        return [ControlLoc(jump_dst)]
+    elif insn.mnemonic == 'jalr':
+        if operands['grs1'] != 1:
+            # This is a function call through a pointer e.g. `jal x1,
+            # <addr>, 0`; not currently supported
+            raise RuntimeError(
+                'Cannot create control graph because of a function '
+                'call from a pointer at PC 0x{:x}: {}\nThis is '
+                'permitted by OTBN but not supported by this check.'.format(
+                    pc, insn.disassemble(pc, operands)))
+        # This jump returns to whatever's on top of the call stack
+        return [Ret()]
```

Comment:
I think you might need to check that the offset is zero too? Otherwise, you won't spot that we're doing something naughty with e.g. `jalr x0, x1, 16`.